### PR TITLE
RFC: use limit true IOContext when showing errors in the REPL

### LINKED ISSUE #14734: BoundsError message contains the entire index, and it can be huge
For BoundsError, the entire index is printed even if it is a large array. Example:

```
julia> [][[true for i=1:100]]
ERROR: BoundsError: attempt to access 0-element Array{Any,1}
  at index [true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true]
 in throw_boundserror at ./abstractarray.jl:147
 [inlined code] from ./abstractarray.jl:154
 in anonymous at ./no file:4294967295
 in eval at ./boot.jl:265
```

I got trapped several times filtering a million-line DataFrame and forgetting the second dimension (like `df[df[:X] .== x]`), but the issue doesn't seem to be specific to DataFrames.

### PR DESCRIPTION
**Pull request by KristofferC**
Makes the example in https://groups.google.com/forum/#!topic/julia-users/ahxFyM082MQ not generate 10 pages of errors in the REPL.

Fixes #14734.

Is this too drastic?

### CODE REVIEW

**Comment by tkelman** [comment]
This should probably be more specific to BoundsError?

**Comment by JeffBezanson** [comment]
I think this is probably ok; limiting output in the repl is the norm, so I don't see a problem with limiting everything inside an error message.

**Comment by KristofferC** [comment]
Anyone else have an opinion here? We could try it for a while and see how people feel.

**Comment by tkelman** [comment]
what outputs is this going to truncate? where `:limit` makes a difference is not something I have much of  an intuition for

**Comment by KristofferC** [comment]
It will truncate in the case that a type specifically does a check for the key `:limit` in the `IOContext` and decides to then print the type a bit more compactly, as in: https://github.com/JuliaLang/julia/blob/06c0ba6233258dbfd9cbb301e3fa115ee1b8de49/base/dict.jl#L53

**Comment by tkelman** [comment]
right, but a showerror function might also choose to print way less under `:limit => true` right? guess not many are doing that, and maybe we should honor it and just be sure no specializations are being so concise as to leave out useful info.

It would be messy global state, but do we have a way of getting the last error to re-show it in full?

**Comment by KristofferC** [comment]
Sure but isn't that likely exactly what the author intended when they chose to limit the output if `:limit => true`? 

Storing the last error message to be able to reshow it / jump to the file location of a stackframe in the editor is something I already do in a package here: https://github.com/KristofferC/OhMyREPL.jl/blob/ea26e96e937ed7c7a4b56f2fb5cb03b23a539102/src/ErrorMessages.jl#L14. It is not really a *messy* global state, not worse than `ans`.

**Comment by JeffBezanson** [comment]
`:limit` refers to truncating large containers; there is a separate flag `:compact` e.g. for printing fewer digits of numbers. But in any case, the intent of `:limit` is just to truncate big containers, not randomly leave out various information.

**Comment by KristofferC** [comment]
Hence, this should be good?

**Comment by tkelman** [comment]
are there any tests for this?

**Comment by KristofferC** [comment]
Test for what?

**Comment by tkelman** [comment]
displaying errors, the behavior change here - it's liable to regress if not tested

**Comment by KristofferC** [comment]
Added test.

File: test/repl.jl [hunk 1]
```diff
@@ -532,3 +532,13 @@ if !is_windows() || Sys.windows_version() >= Sys.WINDOWS_VISTA_VER
     @test readstring(outs) == "1\n"
 end
 end # let exename
+
+# Test containers in error messages are limited #18726
+let io = IOBuffer()
+    REPL.display_error(io,
+        try [][[true for i=1:5000]]
```
Comment (tkelman) [inline test/repl.jl]: aka `trues` ? also indent 2 lines below

Comment (KristofferC) [inline test/repl.jl]: fixed

**Comment by KristofferC** [comment]
Is this OK now?

**Comment by KristofferC** [comment]
bump

**Comment by KristofferC** [comment]
bump

**Comment by mauro3** [comment]
Can this be back-ported?

**Pull request merged**